source/AndroidResolver/src/PlayServicesResolver.cs

```diff
@@ -2191,15 +2192,45 @@ public static void SettingsDialog()
                 // https://docs.gradle.org/3.4/release-notes.html#the-java-library-plugin
                 // https://developer.android.com/studio/releases/gradle-plugin#3-0-0
                 var version = GradleVersion;
+                var versionComparer = new Dependency.VersionComparer();
                 var includeStatement =
                     !String.IsNullOrEmpty(version) &&
-                    (new Dependency.VersionComparer()).Compare("3.4", version) >= 0 ?
+                    versionComparer.Compare("3.4", version) >= 0 ?
                     "implementation" : "compile";
                 if (includeDependenciesBlock) lines.Add("dependencies {");
-                foreach (var packageSpecAndSources in GetPackageSpecs(dependencies: dependencies)) {
-                    lines.Add(String.Format(
-                        "    {0} '{1}' // {2}", includeStatement, packageSpecAndSources.Key,
-                        packageSpecAndSources.Value));
+                // Build a map of dependencies with the max version of that dependency.
+                // If different packages have different versions of the same dependency,
+                // we want to activate only the highest version but still include the
+                // other versions as commented out dependency lines.
+                var dependenciesMaxVersions = new Dictionary<string, string>();
+                foreach( var dependency in dependencies) {
+                    if(!dependenciesMaxVersions.ContainsKey(dependency.VersionlessKey))
+                        dependenciesMaxVersions[dependency.VersionlessKey] = dependency.Version;
+                    else {
+                        var compareWithVersion = dependenciesMaxVersions[dependency.VersionlessKey];
+                        if(versionComparer.Compare(dependency.Version, compareWithVersion) < 0)
+                            dependenciesMaxVersions[dependency.VersionlessKey] = dependency.Version;
+                    }
+                }
+                List<Dependency> dependenciesList = dependencies.OrderBy(Dependency=>Dependency.Key).ToList();
+                foreach( var dependency in dependenciesList) {
+                    // Passing the entire list to GetPackageSpecs returns results
+                    // in an arbitrary order because it takes an IEnumerable.
+                    // Hence we cannot pass the entire list as is. We pass an element
+                    // at a time to ensure the correspondence with results.
+                    List<Dependency> singleItemList = new List<Dependency>();
+                    singleItemList.Add(dependency);
+                    var packageSpecAndSources = GetPackageSpecs(singleItemList)[0];
```

Comment:
Hmm...this is now `GetPackageSpecs()` should be used.

If it is passed with multiple "Dependeny", with the same package name and version, but from different sources, this will combine the sources into a string as the dictionary value.

See this
https://github.com/googlesamples/unity-jar-resolver/blob/30b4f9292db3bd11340883dd80872c9852ae70e3/source/AndroidResolver/src/GradleResolver.cs#L114

After your change, if multiple sources (either through Dependencies.xml or through code) deps on the same package and source, you will see the following in the mainTemplate.gradle.

```
    compile 'com.google.android.gms:play-services-base:15.0.1' // Assets/Editor/FooDependencies.xml:3
    compile 'com.google.android.gms:play-services-base:15.0.1' // Assets/Editor/BarDependencies.xml:3
```

And this is going to cause issue.

You might need to refactor the API a bit to make it work.

Reply:
Great catch and I have refactored the code accordingly. Ended up creating a new function that behaves like `GetPackageSpecs` but returns all other metadata that I require about the version. I am not sure if this is the best way to do it in C# because tuples are only supported in newer versions of .NET. I used something very basic to workaround.   
Side note: It seems like we didn't hit this case of duplicate dependencies because we remove duplicates when scanning xml dependencies. By the time `GradleDependenciesLines` is called, dependencies are already de-duplicated. But your point still stands and is valid as we changed the original behvior.